File: src/Homework1/Homework1Tests/ParserTests.cs
using Homework1;
using Xunit;

namespace Homework1Tests
{
    public class ParserTests
    {
        private const int Correct = 0;
        private const int WrongArgsNumber = 1;
        private const int WrongValueFormat = 2;
        private const int WrongOperation = 3;
        private const int DivisionByZero = 4;
            
        [Theory]
        [InlineData("+", Calculator.SupportedOperations.Plus)]
        [InlineData("-", Calculator.SupportedOperations.Minus)]
        [InlineData("*", Calculator.SupportedOperations.Multiply)]
        [InlineData("/", Calculator.SupportedOperations.Divide)]
        
        public void ParseArgs_Operation_WillParse(string operation, Calculator.SupportedOperations operationExpected)
        {
            var args = new[] { "34", operation, "1253" };
            var parseCode = Parser.ParseArgs(args, out var val1, out var operationResult, out var val2);
            Assert.Equal(Correct, parseCode);
            Assert.Equal(34, val1);
            Assert.Equal(operationExpected, operationResult);
            Assert.Equal(1253, val2);
        }
        [Fact]
        public void ParseCalcArguments_WrongLengthArgs_WillReturnWrongArgsNumber()
        {
            var args = new[] { "98", "+", "1984", "qu" };
            var parseCode = Parser.ParseArgs(args, out _, out _, out _);
            Assert.Equal(WrongArgsNumber, parseCode);
        }
       
        [Fact]
        public void ParseArgs_NotNumber_WillReturnWrongValueFormat()
        {
            var args = new[] { "4", "+", "qu" };
            var parseCode = Parser.ParseArgs(args, out _, out _, out _);
            Assert.Equal(WrongValueFormat, parseCode);
        }
        
        [Fact]
        public void ParseArgss_NotOperation_WillReturnWrongOperation()
        {
            var args = new[] { "4", "qu", "1984" };
            var parseCode = Parser.ParseArgs(args, out _, out _, out _);
            Assert.Equal(WrongOperation, parseCode);
        }

        [Fact]
        public void ParseCalcArguements_DivisionByZero_WillReturnDivisionByZero()
        {
            var args = new[] { "2", "/", "0" };
            var parseCode = Parser.ParseArgs(args, out _, out _, out _);
            Assert.Equal(DivisionByZero, parseCode);
        }
    }
}

File: src/Homework4/Homework4Tests/ParserTests.fs
module Tests.ParserTests

open Homework4
open Xunit

[<Theory>]
[<InlineData("+", Calculator.SupportedOperations.Plus)>]
[<InlineData("-", Calculator.SupportedOperations.Minus)>]
[<InlineData("*", Calculator.SupportedOperations.Multiply)>]
[<InlineData("/", Calculator.SupportedOperations.Divide)>]
let ParseCalcArguments_Operation_WillParse (operation, operationExpected) =
    let args = [|"4";operation;"777"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Calculator.SupportedOperations.Plus
    let check = Parser.parseArgs args &val1 &operationResult &val2
    Assert.Equal(Parser.Parser.Correct, check)
    Assert.Equal(4, val1)
    Assert.Equal(operationExpected, operationResult)
    Assert.Equal(777, val2)

[<Fact>]
let ParseArgs_NotNumber_WillReturnWrongValueFormat () =
    let args = [|"4";"+";"9348hg"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Calculator.SupportedOperations.Plus
    let check = Parser.parseArgs args &val1 &operationResult &val2
    Assert.Equal(Parser.Parser.WrongValueFormat, check)

[<Fact>]
let ParseArgs_NotOperation_WillReturnWrongOperation () =
    let args = [|"4";"qu";"1984"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Calculator.SupportedOperations.Plus
    let check = Parser.parseArgs args &val1 &operationResult &val2
    Assert.Equal(Parser.Parser.WrongOperation, check)

[<Fact>]
let ParseCalcArguments_WrongLengthArgs_WillReturnWrongArgsNumber () =
    let args = [|"4";"+";"1984";"qu"|]
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operationResult = Calculator.SupportedOperations.Plus
    let check = Parser.parseArgs args &val1 &operationResult &val2
    Assert.Equal(Parser.Parser.WrongArgsNumber, check)
    
[<Fact>]
let ParseCalcArguements_DivisionByZero_WillReturnDivisionByZero () =
   let args = [|"2";"/";"0";|]
   let mutable val1 = 0
   let mutable val2 = 0
   let mutable operationResult = Calculator.SupportedOperations.Divide
   let check = Parser.parseArgs args &val1 &operationResult &val2
   Assert.Equal(Parser.Parser.DivisionByZero, check)
     